File: component/component.go
package component

import "github.com/google/uuid"

type Position struct {
	X, Y, Z float32
}

type EntityID uuid.UUID

func NewEntityID() EntityID {
	return EntityID(uuid.New())
}

type Color struct {
	ColorR, ColorG, ColorB float32
	Brightness             float32
}

type TransparentAlpha float32

type Velocity struct {
	DX, DY, DZ float32
}

type RotationVelocity float32

type Angle float32

File: src/component.zig
const std = @import("std");

pub const EntityID = u64;

var next_entity_id: std.atomic.Value(u64) = .init(1);

pub fn newEntityID() EntityID {
    return next_entity_id.fetchAdd(1, .monotonic);
}

pub const Position = extern struct {
    x: f32 = 0,
    y: f32 = 0,
    z: f32 = 0,
};

pub const Velocity = extern struct {
    dx: f32 = 0,
    dy: f32 = 0,
    dz: f32 = 0,
};

pub const Color = extern struct {
    r: f32 = 0,
    g: f32 = 0,
    b: f32 = 0,
    brightness: f32 = 1,
};

pub const TransparentAlpha = f32;
pub const Angle = f32;
pub const RotationVelocity = f32;

test "fresh entity ids are unique and monotonic" {
    const a = newEntityID();
    const b = newEntityID();
    try std.testing.expect(b > a);
}
